### PR TITLE
Add List and NEL .lookup

### LINKED ISSUE #229: List.nth or similar
Is there a method to get the _nth_ element of a `List`, by index? 

If not, would it be welcome or was this omitted on purpose?

### PR DESCRIPTION
**Pull request by mlrv**
Add `List.lookup(i)` and `NEL.lookup(i)` for safe access at index, should be good enough to close #229.

Note that I decided to treat `undefined `values as missing, not sure whether that's the right decision or not, happy to hear feedback on this.

### CODE REVIEW

File: src/monet.js [hunk 1]
```diff
@@ -446,6 +446,9 @@
         headMaybe: function () {
             return this.isNil ? None() : Some(this.head_)
         },
+        lookup: function (i) {
+            return this.isNil ? None() : Maybe.fromUndefined(this.toArray()[i])
```
Comment (ulfryk) [inline src/monet.js:450]: 1. for performance issues I'd recommend checking if `i` is not equal or higher than length
2. `fromUndefined` will throw if value is `null`
Screenshot 2021-03-03 at 8 59 10 am

File: src/monet.js [hunk 2]
```diff
@@ -565,6 +568,10 @@
             return this.head_
         },
 
+        lookup: function (i) {
+            return Maybe.fromUndefined(this.toArray()[i])
```
Comment (ulfryk) [inline src/monet.js:572]: same here

**Review by ulfryk — COMMENTED** [review]
@mlrv looks good - I would just update a little logic to prevent unpleasent exceptions on `null` values.

**Comment by ulfryk** [comment]
@mlrv I'm not sure if it's worth investing time as for now there is decision to drop support for `List` and `NEL` in version `1.0.0` in favour of some small library allowing integration with https://github.com/immutable-js/immutable-js collections.

**Comment by ulfryk** [comment]
But anywy - it's good to have it for now. And It would be great fi you can also add `nth()` that would return value as is ( `undefined` if it's `>=` than length ). That gives end user enough flexibility :)

File: src/monet.js (hunk 1 above)
Comment (mlrv) [reply src/monet.js:450]: > for performance issues I'd recommend checking if i is not equal or higher than length
Yea I guess that makes sense 👍 

> fromUndefined will throw if value is null
Ah got it, thanks

**Comment by mlrv** [comment]
> But anywy - it's good to have it for now. And It would be great fi you can also add `nth()` that would return value as is ( `undefined` if it's `>=` than length ). That gives end user enough flexibility :)

Ah sure, makes sense 👍  It's not a lot of work, so it shouldn't be a big deal if this stuff gets removed in a later release.

In terms of `nth`, I'm not really sure, isn't that against the fundamental principles of this (or any other FP) library? It's trivial to implement of course, but I wonder if it's not just a way to help developers shoot themselves in the foot. Happy to be contradicted of course :)

**Comment by ulfryk** [comment]
> I'm not really sure, isn't that against the fundamental principles of this (or any other FP) library?

I think it's good to allow users to `opt-out` of wrapping `null` / `undefined` elements (this is a cost of a decision to not allow `Some` to contain `null | undefined` 🙃 ).

**Comment by mlrv** [comment]
Mmmh I can see the CI failing, makes sense. Should we rather have `nth` return `null`? Not that it makes much difference, as long as it's documented, but given the nature of the function and how close it is to just a normal index access, I would expect it to return `undefined` in those circumstances

**Comment by ulfryk** [comment]
Yes, undefined is ok.
I believe you can suppress eslint for those 2 lines, as it's a valid place to ignore the rules :)

**Review by ulfryk — APPROVED** [review]

**Pull request merged**